lib/vanagon/component/dsl.rb

```diff
@@ -289,18 +289,56 @@ def environment(env)
         @component.environment.merge!(env)
       end
 
-      # Adds actions to run at the beginning of a package install
+      # Adds action to run during the preinstall phase of packaging
       #
-      # @param action [String] Bourne shell compatible scriptlets to execute
-      def add_preinstall_action(action)
-        @component.preinstall_actions << action
+      # @param pkg_state [Array] the state in which the scripts should execute. Can be
+      #   one or multiple of 'install' and 'upgrade'.
+      # @param block [Array] the Bourne shell compatible scriptlet(s) to execute
+      def add_preinstall_action(pkg_state, block)
+        pkg_state = Array(pkg_state)
+        block = Array(block)
+
+        raise Vanagon::Error, "#{pkg_state} should be an array containing one or more of ['install', 'upgrade']" unless !pkg_state.empty? && (pkg_state - ["install", "upgrade"]).empty?
```

Comment:
This is hard to read as one long line, and with unless immediately followed by `!`. I'd recommend trying if instead of unless and moving the raise inside of an if conditional.